### PR TITLE
refactor(arguments): allow to specify which arguments should be used for every project creation command

### PR DESCRIPTION
**Pull request by DrRaider**
### General

- [ ] Write an explicit title for the Pull Request, following [Conventional Commits specification](https://www.conventionalcommits.org)
- [ ] Test manually the implemented changes
- [ ] Validate the code quality (indentation, syntax, style, simplicity, readability)

### Security

- [ ] Consider the security impact of the changes made

### CODE REVIEW

File: src/abstract-project-create-command.ts [hunk 1]
```diff
@@ -140,69 +119,119 @@ export default abstract class AbstractProjectCreateCommand extends AbstractAuthe
...
+        dbConfig,
+        appConfig,
+        forestAuthSecret: authSecret as string,
+        forestEnvSecret: envSecret as string,
+      });
+
+      await this.notifySuccess();
+    } catch (error) {
+      // Display customized error for non-authentication errors.
+      if (error.status !== 401 && error.status !== 403) {
```
Comment (SteveBunlon) [inline src/abstract-project-create-command.ts:163]: Knowing we are in a `runAuthenticated` method, checking if the error is a 401 one is incorrect to me. At this stage, we are absolutely that the user is authenticated. So only 403 (authorization error) should and can occure here.

File: src/commands/projects/create.ts [hunk 2]
```diff
@@ -1,34 +1,96 @@
...
   private readonly databaseAnalyzer: DatabaseAnalyzer;
 
   private readonly dumper: ForestExpress;
 
+  private readonly commandGenerateConfigGetter: CommandGenerateConfigGetter;
+
+  private readonly _agent: string | null = null;
+
+  static override readonly flags = {
+    ...AbstractProjectCreateCommand.flags,
```
Comment (SteveBunlon) [inline src/commands/projects/create.ts:29]: This feels strange to be honest. On the mother class, it is not specified that flags should be overridden in any place. So, no contracts are specified anywhere that flags should be on the child class. This means that you will rely on the developper comprehension, where OOP is all about specifying contracts that developper should respect. Plus, the child class is responsible for getting back the mother flags. In OOP, the mother class should have this ownership.

File: src/abstract-authenticated-command.ts [hunk 3]
```diff
@@ -15,7 +15,18 @@ export default abstract class AbstractAuthenticatedCommand extends AbstractComma
     this.authenticator = authenticator;
   }
 
-  async checkAuthentication() {
+  async run(): Promise<void> {
```
Comment (SteveBunlon) [inline src/abstract-authenticated-command.ts:18]: Is this really related to the flags?

File: src/abstract-project-create-command.ts [hunk 4]
```diff
@@ -140,69 +119,119 @@ export default abstract class AbstractProjectCreateCommand extends AbstractAuthe
...
+        this.logger.error(['Cannot generate your project.', `${this.messages.ERROR_UNEXPECTED}`]);
+        this.logger.log(`${this.chalk.red(error)}`);
+        this.exit(1);
+      } else {
+        throw error;
+      }
+    }
+  }
+
+  protected abstract processArguments(programArguments: { [name: string]: any }): Promise<{
```
Comment (SteveBunlon) [inline src/abstract-project-create-command.ts:173]: This function name is not self explanatory...

I would rather call it getConfigFormArguments

File: src/interfaces/project-create-interface.ts [hunk 5]
```diff
@@ -40,3 +40,38 @@ export interface Config {
   forestAuthSecret: string;
   forestEnvSecret: string;
 }
+
+interface DbProcessedArgumentsBase {
```
Comment (SteveBunlon) [inline src/interfaces/project-create-interface.ts:44]: Again, processed does not mean anything to me. This looks like command arguments, so this should be called something like `createCommandDbArguments`.

All the stuff you added here is not related to project creation, but command argument parsing. This should reside in a different file, something like `command-create-project-arguments`

File: src/services/dumpers/forest-express.js [hunk 6]
```diff
@@ -46,7 +46,6 @@ class ForestExpress extends AbstractDumper {
     this.isDatabaseLocal = isDatabaseLocal;
     this.toValidPackageName = toValidPackageName;
     this.strings = strings;
-    this.lodash = lodash;
```
Comment (SteveBunlon) [inline src/services/dumpers/forest-express.js:49]: ??

File: src/utils/agents.ts [hunk 7]
```diff
@@ -0,0 +1,7 @@
+enum Agents {
+  NodeJS = 'agent-nodejs',
+  ExpressSequelize = 'express-sequelize',
+  ExpressMongoose = 'express-mongoose',
+}
+
+export default Agents;
```
Comment (SteveBunlon) [inline src/utils/agents.ts:7]:
```suggestion
export default enum Agents {
  NodeJS = 'agent-nodejs',
  ExpressSequelize = 'express-sequelize',
  ExpressMongoose = 'express-mongoose',
}
```

Comment (SteveBunlon) [reply src/utils/agents.ts:7]: Plus, this has nothing to do with flags

File: src/services/prompter/database-prompts.js [hunk 8]
```diff
@@ -25,17 +26,17 @@ class DatabasePrompts extends AbstractPrompter {
...
+    this.handleDialect(forSql, forNosql);
     this.handleName();
     this.handleSchema();
     this.handleHostname();
     this.handlePort();
     this.handleUser();
     this.handlePassword();
     this.handleSsl();
-    this.handleMongodbSrv();
+    if (forNosql) this.handleMongodbSrv();
```
Comment (SteveBunlon) [inline src/services/prompter/database-prompts.js:39]: We already have this kind of checks. What is requested in prompts is handled by the `requests` object. Why didn't you use it to specify the prompts ?

**Review by SteveBunlon — CHANGES_REQUESTED** [review]
I would split it in two. One PR to reorganise the args necessary for a command.

One PR which, based on the new granularity of args, can specify wether or not some options should be requested or not for a command

File: src/abstract-project-create-command.ts (hunk 1 above)
Comment (DrRaider) [reply src/abstract-project-create-command.ts:163]: The token of the user can be discarded during the command execution.
I simply moved this code around, this is not new code.

File: src/commands/projects/create.ts (hunk 2 above)
Comment (DrRaider) [reply src/commands/projects/create.ts:29]: It's the only way to do it. Oclif inheritance is not good.

File: src/abstract-authenticated-command.ts (hunk 3 above)
Comment (DrRaider) [reply src/abstract-authenticated-command.ts:18]: This is just some reordering, I can put it in another PR if you want.

File: src/services/prompter/database-prompts.js (hunk 8 above)
Comment (DrRaider) [reply src/services/prompter/database-prompts.js:39]: Actually it's not realy handled by the `requests` object here, `mongoDbSrv` is always in the `requests` but on the prompt definition we only ask it if the dialect is mongo (so this if is useless you are right)

File: src/interfaces/project-create-interface.ts (hunk 5 above)
Comment (DrRaider) [reply src/interfaces/project-create-interface.ts:44]: 👍

File: src/abstract-project-create-command.ts (hunk 4 above)
Comment (DrRaider) [reply src/abstract-project-create-command.ts:173]: 👍

File: src/abstract-project-create-command.ts (hunk 1 above)
Comment (SteveBunlon) [reply src/abstract-project-create-command.ts]: 👌

File: src/commands/projects/create.ts (hunk 2 above)
Comment (SteveBunlon) [reply src/commands/projects/create.ts:26]: 👌

**Review by SteveBunlon — APPROVED** [review]
